### PR TITLE
Issue #242

### PR DESCRIPTION
**Pull request by daminichopra**
## Proposed changes

Added `backoff_factor` which is used for retry interval between attempt of consecutive next try and so on

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ x ] Bugfix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [ x ] I have read the [CONTRIBUTING][contrib] doc
- [ x ] I have signed the [CLA][cla]
- [ ] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

Refer [Issue](https://github.com/orchestracities/ngsi-timeseries-api/issues/242)

[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

File: docs/manuals/user/index.md [hunk 1]
```diff
@@ -1,456 +1,324 @@
...
-  (See [PR 191](https://github.com/orchestracities/ngsi-timeseries-api/pull/191)),
-  The following limitations still apply: a error in a single data entity will invalidate
-  the all set. There is not optimisation for large message size.
-
-- Data are assumed to be consistent. I.e., if the first data notification for
-  an entity type use a given set of data types for the attributes, the following
-  data notifications must be consistent, or they will be rejected. E.g. if
-  the data type of attribute `speed` of entity type `car` is set initially
-  to `Number`, later on it cannot be set to `Text`.
+# QuantumLeap
```
Comment (chicco785) [inline docs/manuals/user/index.md:1]: this is not part of the pr, remove  the related commits

File: src/reporter/delete.py [hunk 2]
```diff
@@ -34,11 +34,19 @@ def delete_entity(entity_id, type_=None, from_date=None, to_date=None):
 def delete_entities(entity_type, from_date=None, to_date=None,
                     drop_table=False):
     with translator_for(fiware_s()) as trans:
-        if drop_table:
-            trans.drop_table(etype=entity_type, fiware_service=fiware_s())
-            logging.getLogger(__name__).info(
-                "dropped entity_type {}".format(entity_type))
-            return 'entity table dropped', 204
+        if fiware_sp() is not None:
```
Comment (chicco785) [inline src/reporter/delete.py:37]: this is not part of the pr, remove  the related commits

File: src/translators/crate.py [hunk 3]
```diff
@@ -56,7 +56,8 @@ def setup(self):
         self.connection = self.ccm.get_connection('crate')
         if self.connection is None:
             try:
-                self.connection = client.connect([url], error_trace=True)
+                # Added backoff_factor for retry interval between attempt of consecutive retries
+                self.connection = client.connect([url], error_trace=True, backoff_factor=0.1)
```
Comment (chicco785) [inline src/translators/crate.py:60]: @c0c0n3 have a look here, i am not sure what's the impact on queue workers.
also shouldn't this be configurable?

**Review by chicco785 — CHANGES_REQUESTED** [review]

File: src/translators/crate.py (hunk 3 above)
Comment (daminichopra) [reply src/translators/crate.py:60]: > @c0c0n3 have a look here, i am not sure what's the impact on queue workers.
> also shouldn't this be configurable?

@chicco785 , i can add `backoff_factor` as environment variable. If it's ok with you then i can modify code?

Comment (c0c0n3) [reply src/translators/crate.py:60]: > have a look here, i am not sure what's the impact on queue workers.

It should be okay, but there must be a way to configure it since w/ WQ you also have retry intervals, so the admin should be able to configure the Crate client's backoff factor to make sure the total amount spent by the client retrying a connection is less than the minimum retry interval in WQ...

> i can add backoff_factor as environment variable.

yes please!

Comment (chicco785) [reply src/translators/crate.py:60]: @c0c0n3 is there away to check this?

Comment (chicco785) [reply src/translators/crate.py:60]: i meant more an automatic check at start that verify that the back off is smaller than retry interval (which i am not sure how it is set)

File: src/translators/crate.py [hunk 4]
```diff
@@ -57,8 +57,11 @@ def setup(self):
         if self.connection is None:
             try:
                 # Added backoff_factor for retry interval between attempt of consecutive retries
-                self.connection = client.connect([url], error_trace=True, backoff_factor=0.1)
-                self.ccm.set_connection('crate', self.connection)
+                backoff_factor = 0.0
+                if os.environ.get("backoff_factor"):
+                    backoff_factor = float(os.environ["backoff_factor"])
+                    self.connection = client.connect([url], error_trace=True, backoff_factor=backoff_factor)
```
Comment (chicco785) [inline src/translators/crate.py:63]: this line and the following should be outside the if :)

File: src/translators/crate.py [hunk 5]
```diff
@@ -57,8 +57,11 @@ def setup(self):
         if self.connection is None:
             try:
                 # Added backoff_factor for retry interval between attempt of consecutive retries
-                self.connection = client.connect([url], error_trace=True, backoff_factor=0.1)
-                self.ccm.set_connection('crate', self.connection)
+                backoff_factor = 0.0
+                if os.environ.get("backoff_factor"):
```
Comment (chicco785) [inline src/translators/crate.py:61]: this is crate specific, so better call it `CRATE_BACKOFF_FACTOR`

Comment (chicco785) [reply src/translators/crate.py:61]: this should be added in the documentation in the file configuration.md

Comment (chicco785) [reply src/translators/crate.py:61]: @daminichopra the branch is now clean, if you fix my comments above (after synching your local branch with the new remote!!!), we can then merge

Comment (chicco785) [reply src/translators/crate.py:61]: > this is crate specific, so better call it `CRATE_BACKOFF_FACTOR`

@daminichopra I think you missed this comment

File: src/translators/crate.py (hunk 4 above)
Comment (chicco785) [reply src/translators/crate.py:63]: @daminichopra I think you missed this comment

File: src/translators/crate.py (hunk 5 above)
Comment (daminichopra) [reply src/translators/crate.py:61]: > this is crate specific, so better call it `CRATE_BACKOFF_FACTOR`

@chicco785 , backoff_factor is used in crate [code](https://github.com/crate/crate-python/blob/4bf932ace34fd4963f26bccccbedc4ac74ab3282/src/crate/client/http.py#L330)  which is being used in crate.py.
If i used `crate_backoff_factor` it will not recognize this keyword 
Also, let me knowif i missed something. Thanks

Comment (chicco785) [reply src/translators/crate.py:61]: sure, but this is unrelated to the name you give to the ENV VARIABLE :)

File: src/translators/crate.py [hunk 6]
```diff
@@ -54,9 +54,12 @@ def setup(self):
         url = "{}:{}".format(self.host, self.port)
         self.ccm = ConnectionManager()
         self.connection = self.ccm.get_connection('crate')
+        # Added backoff_factor for retry interval between attempt of consecutive retries
+        backoff_factor = EnvReader(log=logging.getLogger(__name__).debug) \
+            .read(FloatVar('CRATE_BACKOFF_FACTOR', 0.0))
         if self.connection is None:
```
Comment (chicco785) [inline src/translators/crate.py:60]: @daminichopra I refactored a bit your code to use facilities included in `utils.cfgreader`. keep this library in mind in case in future you need to read env variables.

**Review by chicco785 — APPROVED** [review]
